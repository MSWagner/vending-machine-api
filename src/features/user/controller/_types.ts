import { ApiProperty } from "@nestjs/swagger";

export class UserPostResponse {
    @ApiProperty()
    readonly tokenType: string;

    @ApiProperty()
    readonly expiresIn: number;

    @ApiProperty({ format: "uuid" })
    refreshToken: string;

    @ApiProperty({ format: "uuid" })
    accessToken: string;
}

export class UserGetResponse {
    @ApiProperty()
    readonly username: string;

    @ApiProperty()
    readonly role: string;

    @ApiProperty()
    readonly deposit?: number;
}

export class SuccessResponse {
    @ApiProperty()
    readonly success: boolean;
}
