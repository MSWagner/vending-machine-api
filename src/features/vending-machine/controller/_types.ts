import { IsEnum, IsInt, IsUUID, Min } from "class-validator";
import { ApiProperty } from "@nestjs/swagger";
import { IPublicProductData } from "src/entities/Product.entity";

export enum ValidCoins {
    Five = 5,
    Ten = 10,
    Twenty = 20,
    Fifty = 50,
    Hundred = 100
}

export const VALID_COINS: number[] = [
    ValidCoins.Hundred,
    ValidCoins.Fifty,
    ValidCoins.Twenty,
    ValidCoins.Ten,
    ValidCoins.Five
];

export class DepositBody {
    @IsInt()
    @IsEnum(ValidCoins)
    @ApiProperty({ type: "int", enum: ValidCoins })
    readonly coin: number;
}

export class BuyBody {
    @IsUUID()
    @ApiProperty({ type: "uuid" })
    readonly productUid: string;

    @IsInt()
    @Min(1)
    @ApiProperty({ type: "int" })
    readonly amount: number;
}

export class BuyResponse {
    @ApiProperty({ type: "product" })
    readonly product: IPublicProductData;

    @ApiProperty({ type: "int" })
    readonly spent: number;

    @ApiProperty({ type: "int", isArray: true })
    readonly change: number[];
}

export class ResetResponse {
    @ApiProperty({ type: "int", isArray: true })
    readonly change: number[];
}

export class SuccessResponse {
    @ApiProperty({ type: "boolean" })
    readonly success: boolean;
}
