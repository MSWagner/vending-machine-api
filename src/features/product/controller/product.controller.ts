import _ from "lodash";
import {
    Controller,
    Request,
    Post,
    Body,
    Param,
    Delete,
    Put,
    UseGuards,
    Get,
    HttpException,
    HttpStatus
} from "@nestjs/common";
import { AuthGuard } from "@nestjs/passport";
import { ApiCreatedResponse, ApiTags } from "@nestjs/swagger";

import { Roles } from "../../auth/guards/role.decorator";
import { IPublicProductData, Product, Role } from "../../../entities/Product.entity";
import { ProductService } from "../services/product.service";
import { RoleGuard } from "../../auth/guards/role.guard";
import { ProductUpsertDto, ProductResponseDto } from "../dto/product.dto";
import { SuccessResponse } from "./_types";

@ApiTags("products")
@Controller({ path: "products", version: "1" })
export class ProductController {
    constructor(private readonly productService: ProductService) {}

    @Post()
    @UseGuards(AuthGuard("bearer"), RoleGuard)
    @Roles(Role.Seller)
    @ApiCreatedResponse({ description: "The product has been successfully created.", type: ProductResponseDto })
    async create(@Request() req, @Body() productDto: ProductUpsertDto): Promise<IPublicProductData> {
        return (await this.productService.create(req.user, productDto)).publicData;
    }

    @Put(":uid")
    @UseGuards(AuthGuard("bearer"), RoleGuard)
    @Roles(Role.Seller)
    @ApiCreatedResponse({ description: "The product has been successfully updated.", type: ProductResponseDto })
    async update(
        @Request() req,
        @Param("uid") uid: string,
        @Body() productDto: ProductUpsertDto
    ): Promise<IPublicProductData> {
        const product = await this.productService.findOne(uid);

        this.checkValidProductOwner(product, req.user.uid);

        return (await this.productService.update(product, req.user, productDto)).publicData;
    }

    @Get()
    @ApiCreatedResponse({ type: ProductResponseDto, isArray: true })
    async findAll(): Promise<IPublicProductData[]> {
        return (await this.productService.findAll()).map((product) => product.publicData);
    }

    @Get(":uid")
    @ApiCreatedResponse({ type: ProductResponseDto })
    async findOne(@Param("uid") uid: string): Promise<IPublicProductData> {
        const product = await this.productService.findOne(uid);

        if (_.isNil(product)) {
            throw new HttpException("Product could not be found.", HttpStatus.NOT_FOUND);
        } else {
            return product.publicData;
        }
    }

    @Delete(":uid")
    @UseGuards(AuthGuard("bearer"), RoleGuard)
    @Roles(Role.Seller)
    @ApiCreatedResponse({ description: "The product has been successfully deleted.", type: SuccessResponse })
    async remove(@Request() req, @Param("uid") uid: string): Promise<SuccessResponse> {
        const product = await this.productService.findOne(uid);

        this.checkValidProductOwner(product, req.user.uid);

        await this.productService.delete(product);

        return { success: true };
    }

    private checkValidProductOwner(product: Product, userUid: string): void {
        if (_.isNil(product)) {
            throw new HttpException("Product could not be found.", HttpStatus.NOT_FOUND);
        }

        if (product.seller.uid !== userUid) {
            throw new HttpException("You are not the owner of this product.", HttpStatus.UNAUTHORIZED);
        }
    }
}
