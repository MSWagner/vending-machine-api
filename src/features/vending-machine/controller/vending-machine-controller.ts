import { Controller, Request, Post, Body, UseGuards, HttpException, HttpStatus, HttpCode } from "@nestjs/common";
import { AuthGuard } from "@nestjs/passport";
import { ApiCreatedResponse, ApiTags } from "@nestjs/swagger";
import { getConnection } from "typeorm";

import { Roles } from "../../auth/guards/role.decorator";
import { Product, Role } from "../../../entities/Product.entity";
import { RoleGuard } from "../../auth/guards/role.guard";

import { User } from "../../../entities/User.entity";

import { BuyBody, BuyResponse, DepositBody, ResetResponse, SuccessResponse, VALID_COINS } from "./_types";

@ApiTags("vending-machine")
@Controller({ version: "1" })
export class VendigMachineController {
    @Post("deposit")
    @HttpCode(HttpStatus.OK)
    @UseGuards(AuthGuard("bearer"), RoleGuard)
    @Roles(Role.Buyer)
    @ApiCreatedResponse({ description: "The buyer has successful deposit coins.", type: SuccessResponse })
    async deposit(@Request() req, @Body() depositBody: DepositBody): Promise<SuccessResponse> {
        const queryRunner = await getConnection().createQueryRunner();
        await queryRunner.startTransaction();

        const user = await queryRunner.manager.getRepository(User).findOne(req.user.uid);
        user.deposit += depositBody.coin;

        try {
            await await queryRunner.manager.getRepository(User).save(user);
            await queryRunner.commitTransaction();
        } catch (err) {
            await queryRunner.rollbackTransaction();
            throw new HttpException(err.message, HttpStatus.INTERNAL_SERVER_ERROR);
        } finally {
            await queryRunner.release();
        }

        return { success: true };
    }

    @Post("buy")
    @HttpCode(HttpStatus.OK)
    @UseGuards(AuthGuard("bearer"), RoleGuard)
    @Roles(Role.Buyer)
    @ApiCreatedResponse({ description: "The buyer has successful bought a product.", type: BuyResponse })
    async buy(@Request() req, @Body() buyBody: BuyBody): Promise<BuyResponse> {
        const queryRunner = await getConnection().createQueryRunner();
        await queryRunner.startTransaction();

        let product = await queryRunner.manager.getRepository(Product).findOne(buyBody.productUid);

        if (!product) {
            throw new HttpException("Product not found", HttpStatus.NOT_FOUND);
        }

        if (product.amountAvailable < 1) {
            throw new HttpException("Product is out of stock", HttpStatus.BAD_REQUEST);
        }

        if (product.amountAvailable < buyBody.amount) {
            throw new HttpException("Not enough products available", HttpStatus.BAD_REQUEST);
        }

        let user = await queryRunner.manager.getRepository(User).findOne(req.user.uid);
        const productsPrice = product.cost * buyBody.amount;

        if (user.deposit < productsPrice) {
            throw new HttpException("Insufficient deposit", HttpStatus.BAD_REQUEST);
        }

        let change = 0;

        try {
            change = user.deposit - productsPrice;
            product.amountAvailable -= buyBody.amount;
            user.deposit = 0;

            product = await queryRunner.manager.getRepository(Product).save(product);
            user = await queryRunner.manager.getRepository(User).save(user);

            await queryRunner.commitTransaction();
        } catch (err) {
            await queryRunner.rollbackTransaction();
            throw new HttpException(err.message, HttpStatus.INTERNAL_SERVER_ERROR);
        } finally {
            await queryRunner.release();
        }

        let changeCoins: number[] = [];
        if (change > 0) {
            changeCoins = this.getChangeCoins(change, VALID_COINS);
        }

        return {
            product: product.publicData,
            spent: productsPrice,
            change: changeCoins
        };
    }

    @Post("reset")
    @HttpCode(HttpStatus.OK)
    @UseGuards(AuthGuard("bearer"), RoleGuard)
    @Roles(Role.Buyer)
    @ApiCreatedResponse({ description: "The buyer has successful reset there deposit.", type: ResetResponse })
    async reset(@Request() req): Promise<ResetResponse> {
        const queryRunner = await getConnection().createQueryRunner();
        await queryRunner.startTransaction();

        const user = await queryRunner.manager.getRepository(User).findOne(req.user.uid);
        const changeCoins = this.getChangeCoins(user.deposit, VALID_COINS);
        user.deposit = 0;

        try {
            await await queryRunner.manager.getRepository(User).save(user);
            await queryRunner.commitTransaction();
        } catch (err) {
            await queryRunner.rollbackTransaction();
            throw new HttpException(err.message, HttpStatus.INTERNAL_SERVER_ERROR);
        } finally {
            await queryRunner.release();
        }

        return { change: changeCoins };
    }

    getChangeCoins(change: number, validCoins: number[]): number[] {
        const changeCoins: number[] = [];

        validCoins
            .sort((a, b) => b - a)
            .forEach((coin) => {
                while (change >= coin) {
                    changeCoins.push(coin);
                    change -= coin;
                }
            });

        return changeCoins;
    }
}
