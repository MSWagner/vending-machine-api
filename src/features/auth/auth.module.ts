import { Module } from "@nestjs/common";

import { AuthService } from "./services/auth.service";

import { LocalStrategy } from "./strategies/local.strategy";
import { BearerStrategy } from "./strategies/bearer.strategy";

import { UserModule } from "../user/user.module";

import { PassportModule } from "@nestjs/passport";

import { accessTokenProviders } from "./providers/accessTokenProviders";
import { refreshTokenProviders } from "./providers/refreshTokenProviders";
import { LoginController } from "./controller/login/login.controller";
import { RefreshTokenController } from "./controller/refresh-token/refresh-token.controller";

@Module({
    imports: [UserModule, PassportModule],
    providers: [AuthService, LocalStrategy, BearerStrategy, ...accessTokenProviders, ...refreshTokenProviders],
    controllers: [LoginController, RefreshTokenController],
    exports: [AuthService]
})
export class AuthModule {}
