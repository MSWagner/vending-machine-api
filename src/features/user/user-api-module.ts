import { Module } from "@nestjs/common";
import { UserController } from "./controller/user.controller";
import { AuthModule } from "../auth/auth.module";
import { UserModule } from "./user.module";

@Module({
    imports: [AuthModule, UserModule],
    controllers: [UserController]
})
export class UserApiModule {}
