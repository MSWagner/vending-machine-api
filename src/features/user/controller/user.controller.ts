import { Controller, Body, Post, HttpException, HttpStatus } from "@nestjs/common";
import { ApiTags, ApiCreatedResponse } from "@nestjs/swagger";

import { AuthService } from "../../auth/services/auth.service";
import { UserService } from "../services/user.service";

import { CONFIG } from "../../../config";

import { UserDto, UserPostResponse } from "./_types";

@ApiTags("user")
@Controller({ path: "user", version: "1" })
export class UserController {
    constructor(private readonly authService: AuthService, private readonly userService: UserService) {}

    @Post()
    @ApiCreatedResponse({ description: "The user has been successfully created.", type: UserPostResponse })
    async handler(@Body() requestDto: UserDto): Promise<UserPostResponse> {
        try {
            const newUser = await this.userService.createUser(
                requestDto.username,
                requestDto.password,
                requestDto.role
            );
            const tokens = await this.authService.generateToken(newUser);

            return {
                tokenType: "Bearer",
                expiresIn: CONFIG.auth.tokenValidity,
                accessToken: tokens.accessToken.token,
                refreshToken: tokens.refreshToken.token
            };
        } catch (err) {
            throw new HttpException("User with the given username already exists", HttpStatus.CONFLICT);
        }
    }
}
