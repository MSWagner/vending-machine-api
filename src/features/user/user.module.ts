import { Module } from "@nestjs/common";
import { UserService } from "./services/user.service";
import { userProviders } from "./user.providers";

import { DatabaseModule } from "../../shared/database/database.module";

@Module({
    imports: [DatabaseModule],
    providers: [...userProviders, UserService],
    exports: [DatabaseModule, UserService, ...userProviders]
})
export class UserModule {}
