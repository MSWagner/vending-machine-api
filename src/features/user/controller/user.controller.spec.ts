import request from "supertest";
import { Test, TestingModule } from "@nestjs/testing";
import { INestApplication } from "@nestjs/common";

import { setupApp } from "../../../appSetup";

import { TestModule } from "../../../shared/test/test.module";
import { TestService } from "../../../shared/test/test.service";
import { UserController } from "./user.controller";
import { UserModule } from "../user.module";
import { AuthModule } from "../../../features/auth/auth.module";

import { User } from "../../../entities/User.entity";

import * as fixtures from "../../../shared/test/fixtures";

describe("User Controller", () => {
    let app: INestApplication;
    let testService: TestService;

    beforeAll(async () => {
        const module: TestingModule = await Test.createTestingModule({
            controllers: [UserController],
            imports: [TestModule, AuthModule, UserModule]
        }).compile();

        testService = module.get<TestService>(TestService);

        app = module.createNestApplication();
        await setupApp(app);
        await app.init();
    });

    beforeEach(async () => {
        await testService.reloadFixtures();
    });

    afterAll((done) => {
        testService.connectionManager.connection.close();
        done();
    });

    it("/POST user - should register an new buyer user", async () => {
        const registerBody = {
            username: "test1",
            password: "passwordTest",
            role: "buyer"
        };

        const userCountBefore = await User.count({ username: registerBody.username });
        expect(userCountBefore).toEqual(0);

        const response = await request(app.getHttpServer()).post("/api/v1/user").send(registerBody).expect(201);

        const userCountAfter = await User.count({ username: registerBody.username });
        expect(userCountAfter).toEqual(1);

        const credentials = testService.replaceValues(response.body, ["UUID", "DATE"]);
        expect(credentials).toMatchSnapshot("RegisterCredentials");
    });

    it("/POST user - should throw an error for a username conflict", async () => {
        const registerBody = {
            username: fixtures.user1.username,
            password: "passwordTest",
            role: "buyer"
        };

        const response = await request(app.getHttpServer()).post("/api/v1/user").send(registerBody).expect(409);

        const credentials = testService.replaceValues(response.body, ["UUID", "DATE"]);
        expect(credentials).toMatchSnapshot("RegisterUsernameConflictError");
    });

    it("/POST user - should throw an error for missing username", async () => {
        const registerBody = {
            password: "passwordTest"
        };

        const response = await request(app.getHttpServer()).post("/api/v1/user").send(registerBody).expect(400);

        const credentials = testService.replaceValues(response.body, ["UUID", "DATE"]);
        expect(credentials).toMatchSnapshot("RegisterUsernameMissingError");
    });

    it("/POST user - should throw an error for missing role", async () => {
        const registerBody = {
            username: "test1",
            password: "passwordTest"
        };

        const response = await request(app.getHttpServer()).post("/api/v1/user").send(registerBody).expect(400);

        const credentials = testService.replaceValues(response.body, ["UUID", "DATE"]);
        expect(credentials).toMatchSnapshot("RegisterRoleMissingError");
    });

    it("/POST user - should throw an error for missing password", async () => {
        const registerBody = {
            username: "test1"
        };

        const response = await request(app.getHttpServer()).post("/api/v1/user").send(registerBody).expect(400);

        const credentials = testService.replaceValues(response.body, ["UUID", "DATE"]);
        expect(credentials).toMatchSnapshot("RegisterPasswordMissingError");
    });

    it("/POST user - should throw an error for a to short password", async () => {
        const registerBody = {
            username: "test1",
            password: "passw"
        };

        const response = await request(app.getHttpServer()).post("/api/v1/user").send(registerBody).expect(400);

        const credentials = testService.replaceValues(response.body, ["UUID", "DATE"]);
        expect(credentials).toMatchSnapshot("RegisterPasswordToShortError");
    });

    it("/POST user - should throw an error for a to short username", async () => {
        const registerBody = {
            username: "te",
            password: "password"
        };

        const response = await request(app.getHttpServer()).post("/api/v1/user").send(registerBody).expect(400);

        const credentials = testService.replaceValues(response.body, ["UUID", "DATE"]);
        expect(credentials).toMatchSnapshot("RegisterUsernameToShortError");
    });

    it("/POST user - should throw an error for not supported role", async () => {
        const registerBody = {
            username: "test1",
            password: "passwordTest",
            role: "not-supported"
        };

        const response = await request(app.getHttpServer()).post("/api/v1/user").send(registerBody).expect(400);

        const credentials = testService.replaceValues(response.body, ["UUID", "DATE"]);
        expect(credentials).toMatchSnapshot("RegisterRoleNotSupportedError");
    });
});
