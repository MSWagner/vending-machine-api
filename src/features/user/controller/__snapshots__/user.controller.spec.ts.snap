// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`User Controller /POST user - should register an new buyer user: RegisterCredentials 1`] = `
{
  "accessToken": "accessToken",
  "expiresIn": 86400000,
  "refreshToken": "refreshToken",
  "tokenType": "Bearer",
}
`;

exports[`User Controller /POST user - should throw an error for a to short password: RegisterPasswordToShortError 1`] = `
{
  "error": "Bad Request",
  "message": [
    "password must be longer than or equal to 6 characters",
    "role must be a valid enum value",
    "role must be longer than or equal to 3 characters",
    "role must be a string",
  ],
  "statusCode": 400,
}
`;

exports[`User Controller /POST user - should throw an error for a to short username: RegisterUsernameToShortError 1`] = `
{
  "error": "Bad Request",
  "message": [
    "username must be longer than or equal to 3 characters",
    "role must be a valid enum value",
    "role must be longer than or equal to 3 characters",
    "role must be a string",
  ],
  "statusCode": 400,
}
`;

exports[`User Controller /POST user - should throw an error for a username conflict: RegisterUsernameConflictError 1`] = `
{
  "message": "User with the given username already exists",
  "statusCode": 409,
}
`;

exports[`User Controller /POST user - should throw an error for missing password: RegisterPasswordMissingError 1`] = `
{
  "error": "Bad Request",
  "message": [
    "password must be longer than or equal to 6 characters",
    "password must be a string",
    "role must be a valid enum value",
    "role must be longer than or equal to 3 characters",
    "role must be a string",
  ],
  "statusCode": 400,
}
`;

exports[`User Controller /POST user - should throw an error for missing role: RegisterRoleMissingError 1`] = `
{
  "error": "Bad Request",
  "message": [
    "role must be a valid enum value",
    "role must be longer than or equal to 3 characters",
    "role must be a string",
  ],
  "statusCode": 400,
}
`;

exports[`User Controller /POST user - should throw an error for missing username: RegisterUsernameMissingError 1`] = `
{
  "error": "Bad Request",
  "message": [
    "username must be longer than or equal to 3 characters",
    "username must be a string",
    "role must be a valid enum value",
    "role must be longer than or equal to 3 characters",
    "role must be a string",
  ],
  "statusCode": 400,
}
`;

exports[`User Controller /POST user - should throw an error for not supported role: RegisterRoleNotSupportedError 1`] = `
{
  "error": "Bad Request",
  "message": [
    "role must be a valid enum value",
  ],
  "statusCode": 400,
}
`;
